packages/upward-js/lib/ParameterValidator.js

```diff
@@ -0,0 +1,110 @@
+    },
+    number() {
+        return value => (isNaN(Number(value)) ? 'must be a number' : '');
+    },
+    any() {
+        return () => '';
+    }
+};
+
+class ParameterValidator {
+    constructor(paramTypes) {
+        this.paramTypes = paramTypes;
+        this.paramValidators = Object.entries(paramTypes).map(
+            ([name, descriptor]) => {
+                const typeTestFactory = typeTests[descriptor.type];
+                const args = descriptor[descriptor.type] || [];
+                const test = typeTestFactory(...args);
+                return {
+                    name,
+                    descriptor,
+                    test
+                };
+            }
+        );
+    }
+    beforeResolved(params) {
+        const report = this.validate(
+            params,
+            ({ resolved, param, name, descriptor, test }) => {
+                const errors = [];
+                if (!params.hasOwnProperty(name)) {
+                    if (descriptor.default) {
+                        resolved[name] = descriptor.default;
+                    } else if (descriptor.required) {
+                        errors.push('is a required value');
+                    }
+                }
+                if (descriptor.literal) {
+                    const typeError = test(param);
+                    if (typeError) {
+                        errors.push(
+                            'must be a literal and not a resolver',
+                            typeError
+                        );
+                    } else {
+                        resolved[name] = param;
+                    }
+                }
+                return errors;
+            }
+        );
+        Object.entries(params).forEach(([name, value]) => {
+            if (!this.paramTypes[name]) {
+                report.errors.push({
+                    name,
+                    value: [`unrecognized property ${name} set to ${value}`]
+                });
+            }
+        });
+    }
+    resolveDependencies(params) {
+        return this.validate(params, ({ resolved, param, name, test }) => {
+            if (params.hasOwnProperty(param)) {
+                const typeError = test(param);
+                if (typeError) {
+                    return [typeError];
+                }
+                resolved[name] = param;
+            }
+        });
+    }
+    validate(params, reporter) {
+        const report = {
+            errors: [],
+            resolved: {}
+        };
+        this.paramValidators.forEach(({ name, descriptor, test }) => {
+            const itemErrors = reporter({
+                resolved: report.resolved,
+                param: params[name],
+                name,
+                descriptor,
+                test
+            });
+
+            if (itemErrors.length > 0) {
+                errors.push({
+                    name,
+                    errors: itemErrors
+                });
```

Comment:
It doesn't look like `errors` is defined at this point, so `errors.push` looks wrong to me here. Did you mean `report.errors.push()`?